doc/elements-tx-format.md

```diff
@@ -0,0 +1,562 @@
+# Elements Transaction Format
+
+This document describes the format used to serialize Elements transactions. Once a transaction has been converted into this raw, serialized form, it can be broadcast across the network.
+
+This document assumes some familiarity with Bitcoin and Elements (UTXOs, [Script](https://en.bitcoin.it/wiki/Script), assets, pegins, etc.). For more information on those, please refer to the [Bitcoin Wiki](https://en.bitcoin.it/wiki/Main_Page) and the [Elements Code Tutorial](https://elementsproject.org/elements-code-tutorial/overview).
+
+### Data Types
+
+*Notes*:
+* Fields in the following table are listed in the same order in which they are serialized.
+
+* The *Encoding* column in the following tables indicates how the fields are serialized. All fields in Elements which are the result of a hash (TXID, asset ID, etc.) have their bytes serialized in reverse order. For example: the hash `0x112345...` would be serialized as `...543211`.[^1]
```

Comment:
I don't mean to keep harping on this, and if you want to leave it as it, that is acceptable, but I think this still still all very confusing/misleading.

A hash, as in the output of the SHA-256 function, is **not** being serialized in reverse order.  Instead it is the Element's CLI and GUI interfaces that are displaying these hash values in reverse byte-order.

I get the problem you are having: your main point here is that the order serialized and the order displayed by the Elements UI are reversed from one and other, so who is to say which one is backwards and which one is forward?

But for a developer who is making transactions and computing hashes by hand, there is a clear canonical order: the order produced by ever SHA-256 library out there.  And that canonical order is the order that appears in the serialization format.

Reply:
Agreed -- nothing in the binary serialization is reversed. I think it is confusing to the point of being incorrect to say otherwise.

Reply:
I understand that the SHA-256 algorithm's output corresponds directly to the serialized hash. However, is this what we want to call the "canonical order"?

If someone sends me a payment and I ask them "what is the TXID", should I be expecting them to send me the reverse of what the Elements client will show (i.e. the aforementioned canonical order)? I would personally be expecting to see the same TXID that Elements would show.

I am fine with either way, I just want to confirm that we are all on the same page. Thanks! :) 

(And for what it's worth, I briefly touched on this subtlety in the attached footnote:
> [^1]: This is for historical reasons. In Bitcoin, numbers and hashes are interpreted as being encoded in little-endian, and this same policy has been carried over to Elements. So, when Elements encounters a hash (TXID, Asset ID, etc.) it interprets it as a little-endian integer and parses the bytes backwards.
)

Reply:
To be clear, in my earlier comments suggesting "reversed" language rather than "endianness" language, I missed the context, which is that the binary encoding doesn't actually have any hash reversals in it so there is no need to mention this stuff at all.

Reply:
>I understand that the SHA-256 algorithm's output corresponds directly to the serialized hash. However, is this what we want to call the "canonical order"?

Yes.

>If someone sends me a payment and I ask them "what is the TXID", should I be expecting them to send me the reverse of what the Elements client will show (i.e. the aforementioned canonical order)? I would personally be expecting to see the same TXID that Elements would show.

I would be expecting them to send a hex-encoded byte-reversed version of the txid. It is easy to tell that it is byte-reversed because it is in hex.

Reply:
I agree with you that this is not directly relevant here.

However, a serialized transaction contains outpoint TXIDs, which are reversed when serialized / deserialized. So, if a developer is trying to create or parse a raw transaction, this might be useful for them to understand.

Nevertheless, it is certainly possible to entirely omit any mentions of this, and that might be the best move.

Reply:
Yeah, I think it's reasonable to add a footnote saying that hex encodings of these values are usually reversed, and will need to be re-reversed to match the serialized data. And also mention that this is exactly the same as the situation in Bitcoin.

Reply:
The canonical order of a TXID is the reverse of the canonical order of a SHA-256 hash.  I totally get why this is a mess to explain.

Reply:
I don't like this framing -- the implication is then that the serialized order of a txid (in a transaction input) is reversed.

Reply:
The serialization order is reversed for a TXID.  Certainly when I ask for a TXID what I get is a hex string that is the reverse of what is serialized in the transaction.

Reply:
I think my suggestion is not to call this big or little endian.  There are two orders: hash order and txid/assetid order.  These two orders are reversed from each other.  The format being serialized is hash order, which is the reverse of the txid and assetid orders.

Reply:
I can live with that, although I still find it confusing because it's inconsistent with the way that I think about these objects.

I consider "reversed hex" to be the encoding of these types and still consider the natural order of their bytes to be the ones that come out of the sha2 engine, even for TXIDs/blockhashes/etc (which I agree are distinct types from sha2 hashes, if only because they are encoded differently for user display :)).

Reply:
Update: thank you for your feedback. I have omitted all mentions of this byte-reversal from the body of the document. Instead, in the encoding column for the *Output Hash (TXID)* field, I have linked to a footnote which reads:

> The hex encodings of hashes by the Elements client (TXID, asset ID) are reversed, and so they will need to be re-reversed to match the serialized data. This is the same situation as in Bitcoin. For example, the hash `0x1123...7899` would be displayed by the Bitcoin and Elements clients as `9987...3211`. This is primarily for historical reasons: the Bitcoin client has always interpreted and displayed hashes as little-endian integers and parsed their bytes in reverse order.


In other words, I have explained the situation without identifying which order is the "canonical order'. I have simply stated which order is the hash order, and which order is used for display by the Elements client. The user can interpret it how they wish, so long as they understand that the underlying hash corresponds to the serialized order, and the bytes are reversed for display by the client.

Please let me know if this is still confusing.

Thank you!

Reply:
Marking this resolved to try to convince github not to hide other, active discussions.